### PR TITLE
FullyConnected Bias performance improvement on GPU

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
This PR improves performance of bias addition kernel (both fwd and bwd) in FullyConnected operator.

@eric-haibin-lin FYI

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Cuda, Operator, pr-awaiting-review]

File: src/operator/nn/fully_connected-inl.h [hunk 1]
```diff
@@ -122,10 +171,147 @@ void FCForward(const OpContext &ctx, const FullyConnectedParam &param,
...
+    auto stream = mshadow::Stream<gpu>::GetStream(s);
+    AddBiasGradKernelPhase1<LType><<<n_blocks, 512, 0, stream>>>(scratch_space.dptr_,
+                                                                 grad.dptr_, N, M);
+    AddBiasGradKernelPhase2<<<(N + 127) / 128, 128, 0, stream>>>(scratch_space.dptr_,
+                                                                 gbias.dptr_, N,
+                                                                 blocks_y, req);
+  });
+}
+#endif
+
```
Comment (DickJC123) [inline src/operator/nn/fully_connected-inl.h]: Could you make this a little easier to understand with a ceil_div() macro or inline function, plus eliminate any magic constants, e.g.:
```
#define ceil_div(x, y) (((x) + (y) - 1) / (y))
const unsigned int threads_per_warp 32;
const unsigned int warp_load_bytes = threads_per_warp * sizeof(LType);
const unsigned int blocks_x = ceil_div(N * sizeof(DType), warp_load_bytes);
etc...
```
512 and 128 are also magic constants here.

Comment (ptrendx) [reply src/operator/nn/fully_connected-inl.h]: ok

File: src/operator/nn/fully_connected-inl.h [hunk 2]
```diff
@@ -169,19 +355,7 @@ void FCBackward(const OpContext &ctx, const FullyConnectedParam &param,
...
-    mxnet::TShape small;
-    if (shape_assign(&gbias_blob.shape_, Shape2(param.num_hidden, 1))) {
-      small = gbias_blob.shape_;
-    } else {
-      small = ReduceAxesShapeImpl(grad_blob.shape_, dmlc::optional<mxnet::TShape>(x), true, false);
-    }
-    ReduceAxesComputeImpl<xpu, mshadow::red::sum, false, false,
-                          mshadow_op::identity>(ctx, {grad_blob}, {req[fullc::kBias]},
-                                                {in_grad[fullc::kBias]}, small);
+      AddBiasGrad(in_grad[fullc::kBias], grad, req[fullc::kBias], param.num_hidden, ctx);
```
[the remark this answers is not in the document]
Comment (ptrendx) [reply src/operator/nn/fully_connected-inl.h]: Summary of the time improvement:
For float32
 - FWD: from 396.9 -> 390.7
 - BWD: from 501.9 -> 285.4
For float16:
 - FWD: from 247.1 -> 177.8
 - BWD: from 543.8 -> 203.7

File: src/operator/nn/fully_connected-inl.h [hunk 3]
```diff
@@ -272,20 +284,26 @@ void AddBiasGrad(const TBlob& in_grad,
   int ltype = mxnet::common::cuda::get_load_type(N * sizeof(DType));
   const int M = grad_blob.shape_.Size() / N;
   MXNET_LOAD_TYPE_SWITCH(ltype, LType, {
-    const unsigned int blocks_x = (N * sizeof(DType) + 32 * sizeof(LType) - 1) /
-                                  (32 * sizeof(LType));
+    const unsigned int blocks_x = ceil_div(N * sizeof(DType), 32 * sizeof(LType));
```
Comment (DickJC123) [inline src/operator/nn/fully_connected-inl.h]: Is this 32 the threads_per_warp?  So:
```
const unsigned int blocks_x = ceil_div(N * sizeof(DType), threads_per_warp * sizeof(LType));
```

Comment (ptrendx) [reply src/operator/nn/fully_connected-inl.h]: ok

Comment (DickJC123) [reply src/operator/nn/fully_connected-inl.h]: Very clear now.  LGTM.

**Pull request merged**